### DISCUSSION TITLE
Creating a message and marshaling

### DISCUSSION THREAD

**Discussion by fostercraig** [choria-io/general#37] [Q&A] [answered]
I'm looking in the docs and code and it's not super clear how I can get a Choria FW message with the formatting I expect but with proper JSON marshaling. Am I missing something here? 

If I use [NewMessage](https://github.com/choria-io/go-choria/blob/6ddbb021894cb5dfcacbd1ad1554d4eec6755665/message/message.go#L84-L91), the return type is a [struct `inter.Message` ](https://github.com/choria-io/go-choria/blob/main/inter/message.go#L23)but that function calls `newMessage` which will return `*Message`. I'm not sure if I'm just able to marshal that directly (actually pretty sure not) or if I need to somehow convert the result to an [implementation](https://github.com/choria-io/go-choria/blob/main/message/message.go#L21) first.

**Comment by fostercraig** [comment]
I guess looking in the tests helps:
https://github.com/choria-io/go-choria/blob/6ddbb021894cb5dfcacbd1ad1554d4eec6755665/message/message_test.go#L190

I would suggest that examples be documented though. I can see I need to implement with `*message.Message` but even in the test examples it's not qualified so someone naive like myself is going to need to poke around a bit to understand what's going on. 

The next problem I see is the fields in that struct are not exported, so I won't be able to actually Marshall that. Basically, I want to transfer these Choria messages over NATS JetStream and then decode them as Choria messages with the proper request ID, etc.

**Comment by ripienaar** [comment, marked as answer]
Can you tell me what you are trying to achieve? Generally no need to go that deep for a user.  

You generate a client against a DDL using “choria plugin generate client” and just use that wrapper. You never need to know this stuff. 

A message itself never get serialized it’s used to build the protocol.Request which makes a protocol.SecurRequest which make a protocol.RequestTransport and that’s what’s being serialized, sent et 

I can explain more but would be helpful to know your goal

**Reply by vjanelle** [reply]
We're looking at re-using the message wrappers for our own purposes.

**Reply by ripienaar** [reply]
It’s very inefficient as things go - but anyway you should look here https://github.com/choria-io/go-choria/blob/main/choria/protocol.go for how to go from payload to transport via security and back

**Reply by vjanelle** [reply]
Thanks!

This'll be transporting all of maybe dozens of messages a day - transport integrity and identity is more important than throughput in this particular case.

**Reply by ripienaar** [reply]
Nice. Well I can help you do that it’s quite easy. But you will need to fire up a security provider - easy with a choria framework instance then just call these. 

The core Request / Reply can carry any bytes payload doesn’t need to be rpc

**Reply by fostercraig** [reply]
Sorry, I'm delayed in responding but thanks for the answers. I think I did get something working out with the correct wrappers/headers and now have to create that provider.
